File: src/diff.rs
use serde::{Deserialize, Serialize};
use std::fmt;

/// A single edit operation with byte-offset positions.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Edit {
    Insert {
        pos: usize,
        text: String,
    },
    Delete {
        pos: usize,
        len: usize,
    },
    Replace {
        pos: usize,
        old_len: usize,
        new_text: String,
    },
}

/// An ordered collection of edits with a checksum of the source text they were
/// computed against. Applying the edits to a string matching the checksum
/// produces the target text.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EditList {
    pub edits: Vec<Edit>,
    pub checksum: String,
}

impl EditList {
    pub fn new(edits: Vec<Edit>, source: &str) -> Self {
        Self {
            edits,
            checksum: checksum(source),
        }
    }

    pub fn empty(source: &str) -> Self {
        Self::new(Vec::new(), source)
    }

    pub fn is_empty(&self) -> bool {
        self.edits.is_empty()
    }

    pub fn len(&self) -> usize {
        self.edits.len()
    }
}

impl fmt::Display for EditList {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.edits.is_empty() {
            write!(f, "No edits")
        } else {
            write!(f, "{} edits:", self.edits.len())?;
            for (i, edit) in self.edits.iter().enumerate() {
                write!(f, "\n  {}: {edit:?}", i + 1)?;
            }
            Ok(())
        }
    }
}

/// Simple checksum combining length and char-value sum. Not cryptographic —
/// used only for quick shadow-equality checks during sync.
pub fn checksum(text: &str) -> String {
    let hash = text.len() ^ (text.chars().map(|c| c as u32).sum::<u32>() as usize);
    format!("{hash:x}")
}

/// Compute the minimal edit list to transform `from` into `to`.
///
/// Strips common prefix and suffix, then emits a single Insert, Delete, or
/// Replace for the differing middle section. Positions are byte offsets.
pub fn diff(from: &str, to: &str) -> EditList {
    if from == to {
        return EditList::empty(from);
    }
    if from.is_empty() {
        return EditList::new(
            vec![Edit::Insert {
                pos: 0,
                text: to.to_string(),
            }],
            from,
        );
    }
    if to.is_empty() {
        return EditList::new(
            vec![Edit::Delete {
                pos: 0,
                len: from.len(),
            }],
            from,
        );
    }

    let from_chars: Vec<char> = from.chars().collect();
    let to_chars: Vec<char> = to.chars().collect();

    let mut common_start = 0;
    while common_start < from_chars.len()
        && common_start < to_chars.len()
        && from_chars[common_start] == to_chars[common_start]
    {
        common_start += 1;
    }

    let mut common_end = 0;
    while common_end < from_chars.len() - common_start
        && common_end < to_chars.len() - common_start
        && from_chars[from_chars.len() - 1 - common_end]
            == to_chars[to_chars.len() - 1 - common_end]
    {
        common_end += 1;
    }

    let prefix_bytes: usize = from_chars[..common_start]
        .iter()
        .map(|c| c.len_utf8())
        .sum();
    let suffix_bytes: usize = if common_end > 0 {
        from_chars[from_chars.len() - common_end..]
            .iter()
            .map(|c| c.len_utf8())
            .sum()
    } else {
        0
    };

    let from_middle = &from[prefix_bytes..from.len() - suffix_bytes];
    let to_middle = &to[prefix_bytes..to.len() - suffix_bytes];

    let edit = match (from_middle.is_empty(), to_middle.is_empty()) {
        (true, true) => return EditList::empty(from),
        (true, false) => Edit::Insert {
            pos: prefix_bytes,
            text: to_middle.to_string(),
        },
        (false, true) => Edit::Delete {
            pos: prefix_bytes,
            len: from_middle.len(),
        },
        (false, false) => Edit::Replace {
            pos: prefix_bytes,
            old_len: from_middle.len(),
            new_text: to_middle.to_string(),
        },
    };

    EditList::new(vec![edit], from)
}

/// Apply edits to `text`, returning the transformed result.
///
/// Edits are applied in reverse order to avoid cascading position shifts.
/// Positions are clamped to text bounds for fuzzy-patch tolerance.
pub fn patch(text: &str, edit_list: &EditList) -> Result<String, PatchError> {
    if edit_list.is_empty() {
        return Ok(text.to_string());
    }

    let mut result = text.to_string();

    for edit in edit_list.edits.iter().rev() {
        match edit {
            Edit::Insert {
                pos,
                text: insert_text,
            } => {
                let safe_pos = (*pos).min(result.len());
                result.insert_str(safe_pos, insert_text);
            }
            Edit::Delete { pos, len } => {
                let start = (*pos).min(result.len());
                let end = (start + len).min(result.len());
                if start < end {
                    result.drain(start..end);
                }
            }
            Edit::Replace {
                pos,
                old_len,
                new_text,
            } => {
                let start = (*pos).min(result.len());
                let end = (start + old_len).min(result.len());
                result.replace_range(start..end, new_text);
            }
        }
    }

    Ok(result)
}

/// Errors that can occur during patch application.
#[derive(Debug, Clone, PartialEq)]
pub enum PatchError {
    ChecksumMismatch,
    InvalidPosition,
    InvalidEdit,
}

impl fmt::Display for PatchError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::ChecksumMismatch => write!(f, "Checksum mismatch"),
            Self::InvalidPosition => write!(f, "Invalid position"),
            Self::InvalidEdit => write!(f, "Invalid edit"),
        }
    }
}

impl std::error::Error for PatchError {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_diff_and_patch() {
        let original = "The quick brown fox";
        let modified = "The quick red fox jumps";

        let edits = diff(original, modified);
        let result = patch(original, &edits).unwrap();
        assert_eq!(result, modified);
    }

    #[test]
    fn test_empty_diff() {
        let text = "Same text";
        assert!(diff(text, text).is_empty());
    }

    #[test]
    fn test_fuzzy_patch() {
        let original = "Hello world";
        let modified = "Hello beautiful world";
        let edits = diff(original, modified);

        let different_text = "Hello cruel world";
        let result = patch(different_text, &edits).unwrap();
        assert!(result.contains("beautiful"));
    }
}

File: web/lib/diff.ts
/**
 * Diff and patch engine matching the Rust implementation in src/diff.rs.
 *
 * All positions (pos, len, old_len) are byte offsets — Rust strings are UTF-8
 * byte arrays. We use TextEncoder/TextDecoder to produce compatible offsets
 * so edits can cross the language boundary.
 */

const encoder = new TextEncoder();
const decoder = new TextDecoder();

export type Edit =
  | { Insert: { pos: number; text: string } }
  | { Delete: { pos: number; len: number } }
  | { Replace: { pos: number; old_len: number; new_text: string } };

export interface EditList {
  edits: Edit[];
  checksum: string;
}

/** Simple checksum matching the Rust implementation: byte_len XOR char_code_sum. */
export function checksum(text: string): string {
  let charSum = 0;
  for (const ch of text) {
    charSum += ch.codePointAt(0)!;
  }
  const result = encoder.encode(text).length ^ charSum;
  return (result >>> 0).toString(16);
}

/** Compute the minimal edit list to transform `from` into `to`. */
export function diff(from: string, to: string): EditList {
  if (from === to) {
    return { edits: [], checksum: checksum(from) };
  }

  if (from.length === 0) {
    return {
      edits: [{ Insert: { pos: 0, text: to } }],
      checksum: checksum(from),
    };
  }

  if (to.length === 0) {
    return {
      edits: [{ Delete: { pos: 0, len: encoder.encode(from).length } }],
      checksum: checksum(from),
    };
  }

  const fromChars = Array.from(from);
  const toChars = Array.from(to);

  let commonStart = 0;
  while (
    commonStart < fromChars.length &&
    commonStart < toChars.length &&
    fromChars[commonStart] === toChars[commonStart]
  ) {
    commonStart++;
  }

  let commonEnd = 0;
  while (
    commonEnd < fromChars.length - commonStart &&
    commonEnd < toChars.length - commonStart &&
    fromChars[fromChars.length - 1 - commonEnd] ===
      toChars[toChars.length - 1 - commonEnd]
  ) {
    commonEnd++;
  }

  const prefixBytes = encoder.encode(
    fromChars.slice(0, commonStart).join("")
  ).length;

  const suffixBytes =
    commonEnd > 0
      ? encoder.encode(
          fromChars.slice(fromChars.length - commonEnd).join("")
        ).length
      : 0;

  const fromMiddleBytes =
    encoder.encode(from).length - prefixBytes - suffixBytes;
  const toMiddle = toChars
    .slice(commonStart, toChars.length - commonEnd || undefined)
    .join("");

  const edits: Edit[] = [];

  if (fromMiddleBytes > 0 || toMiddle.length > 0) {
    if (fromMiddleBytes === 0) {
      edits.push({ Insert: { pos: prefixBytes, text: toMiddle } });
    } else if (toMiddle.length === 0) {
      edits.push({ Delete: { pos: prefixBytes, len: fromMiddleBytes } });
    } else {
      edits.push({
        Replace: {
          pos: prefixBytes,
          old_len: fromMiddleBytes,
          new_text: toMiddle,
        },
      });
    }
  }

  return { edits, checksum: checksum(from) };
}

/**
 * Apply edits to `text`, returning the transformed result.
 * Edits are applied in reverse order to avoid cascading position shifts.
 * Positions are clamped to byte-array bounds for fuzzy-patch tolerance.
 */
export function patch(text: string, editList: EditList): string {
  if (editList.edits.length === 0) {
    return text;
  }

  let bytes = encoder.encode(text);

  for (let i = editList.edits.length - 1; i >= 0; i--) {
    const edit = editList.edits[i];

    if ("Insert" in edit) {
      const pos = Math.min(edit.Insert.pos, bytes.length);
      const insertBytes = encoder.encode(edit.Insert.text);
      const result = new Uint8Array(bytes.length + insertBytes.length);
      result.set(bytes.slice(0, pos));
      result.set(insertBytes, pos);
      result.set(bytes.slice(pos), pos + insertBytes.length);
      bytes = result;
    } else if ("Delete" in edit) {
      const start = Math.min(edit.Delete.pos, bytes.length);
      const end = Math.min(start + edit.Delete.len, bytes.length);
      if (start < end) {
        const result = new Uint8Array(bytes.length - (end - start));
        result.set(bytes.slice(0, start));
        result.set(bytes.slice(end), start);
        bytes = result;
      }
    } else if ("Replace" in edit) {
      const start = Math.min(edit.Replace.pos, bytes.length);
      const end = Math.min(start + edit.Replace.old_len, bytes.length);
      const newTextBytes = encoder.encode(edit.Replace.new_text);
      const result = new Uint8Array(
        bytes.length - (end - start) + newTextBytes.length
      );
      result.set(bytes.slice(0, start));
      result.set(newTextBytes, start);
      result.set(bytes.slice(end), start + newTextBytes.length);
      bytes = result;
    }
  }

  return decoder.decode(bytes);
}
